### PR TITLE
handle assumeRoleChain config with length > 1 in no-fork clients

### LINKED ISSUE #998: cannot configure no-fork client because the length of assume role chain array is more than 1
### What happened?

After upgrading the aws provider in a dev environment to test `v0.44.0` we saw a problem with the reconciliation of the MRs deployed:
```
Warning  CannotConnectToProvider  38s (x6 over 101s)      managed/iam.aws.upbound.io/v1beta1, kind=role  cannot initialize the no-fork async external client: cannot get terraform setup: could not configure no-fork AWS client: cannot configure no-fork client because the length of assume role chain array is more than 1
```
A similar behavior was described in https://github.com/upbound/provider-aws/issues/971 and was expected to be resolved in `v0.45.0`. 

However, we see this (other?) issue after upgrading to `v0.45.0`. All the MRs from different split aws providers fail with:
```
    Last Transition Time:  2023-12-01T13:39:31Z
    Message:               connect failed: cannot initialize the no-fork async external client: cannot get terraform setup: could not configure no-fork AWS client: cannot configure no-fork client because the length of assume role chain array is more than 1
    Reason:                ReconcileError
    Status:                False
    Type:                  Synced
```

### How can we reproduce it?

```
❯ kc get providers
NAME                          INSTALLED   HEALTHY   PACKAGE                                                                 AGE
provider-argocd               True        True      xpkg.upbound.io/crossplane-contrib/provider-argocd:v0.1.0-26.ga23f109   65d
provider-aws-autoscaling      True        True      xpkg.upbound.io/upbound/provider-aws-autoscaling:v0.45.0                18d
provider-aws-ec2              True        True      xpkg.upbound.io/upbound/provider-aws-ec2:v0.45.0                        65d
provider-aws-eks              True        True      xpkg.upbound.io/upbound/provider-aws-eks:v0.45.0                        65d
provider-aws-iam              True        True      xpkg.upbound.io/upbound/provider-aws-iam:v0.45.0                        65d
provider-aws-kafka            True        True      xpkg.upbound.io/upbound/provider-aws-kafka:v0.45.0                      22d
provider-kubernetes           True        True      xpkg.upbound.io/crossplane-contrib/provider-kubernetes:v0.9.0           65d
upbound-provider-family-aws   True        True      xpkg.upbound.io/upbound/provider-family-aws:v0.45.0                     6m53s
```
Provider config:
```
API Version:  aws.upbound.io/v1beta1
Kind:         ProviderConfig
...
Spec:
  Assume Role Chain:
    Role ARN:  arn:aws:iam::<ACCOUNT1>:role/service-role/crossplane1
    Role ARN:  arn:aws:iam::<ACCOUNT2>:role/service-role/crossplane2
  Credentials:
    Source:  IRSA
```
### What environment did it happen in?

* Crossplane Version: `v1.13.2`
* Provider Version: `v0.45.0`
* Kubernetes Version: `v1.26.10-eks-4f4795d`  
* Kubernetes Distribution: EKS

### PR DESCRIPTION
**Pull request by erhancagirici**
### Description of your changes

Handles `assumeRoleChain` config with chain length greater than 1, in the AWS provider config. Native terraform provider config block cannot handle this (see https://github.com/hashicorp/terraform-provider-aws/issues/22728), therefore the AssumeRole STS operations are handled in provider code and obtained AWS STS temporary credentials are used.

Fixes #998

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Tested with an AWS ProviderConfig specifying AssumeRoleChain of 3, with cross-account roles assuming each other. A VPC object is successfully created using this provider config.

### CODE REVIEW

File: internal/clients/aws.go [hunk 1]
```diff
@@ -302,43 +295,29 @@ func configureNoForkAWSClient(ctx context.Context, c client.Client, mg resource.
...
-			awsConfig.AssumeRole.ExternalID = *pc.Spec.AssumeRoleChain[0].ExternalID
-		}
-
-		tags := make(map[string]string)
-		for i, tag := range pc.Spec.AssumeRoleChain[0].Tags {
-			if tag.Key == nil || tag.Value == nil {
-				return xpprovider.AWSConfig{}, errors.Errorf("cannot configure no-fork client: tag key or value cannot be nil in spec.AssumeRoleChain[0].Tags[%d]", i)
-			}
-			tags[*tag.Key] = *tag.Value
+	if len(pc.Spec.AssumeRoleChain) != 0 {
```
Comment (ulucinar) [inline internal/clients/aws.go:313]: I think we can call `cfg.Credentials.Retrieve(ctx)` unconditionally (please also see the comment above for the `authKeySecret` case:
```suggestion
```

**Comment by erhancagirici** [comment]
/test-examples="examples/iam/role.yaml"

**Comment by erhancagirici** [comment]
/test-examples="examples/iam/role.yaml"

**Comment by erhancagirici** [comment]
/test-examples="examples/iam/role.yaml"

**Comment by erhancagirici** [comment]
Tested all auth methods, long-term AccessKey/SecretKey from secret, IRSA, WebIdentity on top of IRSA. All of these are combined with AssumeRoleChains too. All of these are passing.

**Review by ulucinar — APPROVED** [review]
Thanks @erhancagirici, lgtm.

**Pull request merged**